use (
  alias://expected
  ConstDatum, DatumError
)

suite(__filename, proc()
  suite("post", proc()
    test("when merge, this must be merged to the value", proc()
      # (1) act
      const out = ConstDatum(name = "c", merge = {y = 2}, value = {x = 1, z = 3})

      # (2) assessment
      expected(out.getValue()).equalTo(x = 1, y = 2, z = 3)
    end)

    test("when append, this must be concatenated to the value", proc()
      # (1) act
      const out = ConstDatum(name = "c", append = [3, 4], value = [1, 2])

      # (2) assessment
      expected(out.getValue([1, 2, 3, 4]))
    end)

    test("when prepend, value must be concatenated to this", proc()
      # (1) act
      const out = ConstDatum(name = "c", prepend = [3, 4], value = [1, 2])

      # (2) assessment
      expected(out.getValue([3, 4, 1, 2]))
    end)
  end)

  suite("setValue()", proc()
    test("when setValue called, error must be raised", proc()
      # (1) act
      const (
        datum = ConstDatum(name = "c", value = 123)
        out = peval(datum.setValue(321))
      )

      # (2) assessment
      expected(out)\
        .it(0).equalTo(false)\
        .it(1).toBe(DatumError).like("Datum 'c' is not updatable.")
    end)
  end)
end)
